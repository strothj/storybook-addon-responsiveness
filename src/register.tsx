import * as React from "react";
import addons from "./addons";
import { addonName, addonPanelName, addonTitle } from "./constants";
import Manager from "./components/Manager";

let registered = false;

export const register = () => {
  if (registered) return;
  registered = true;

  addons.register(addonName, api => {
    addons.addPanel(addonPanelName, {
      title: addonTitle,
      render: () => (
        <Manager storybookChannel={addons.getChannel()} storybookAPI={api} />
      ),
    });
  });
};

export const testingResetRegistered = () => {
  registered = false;
};

register();
