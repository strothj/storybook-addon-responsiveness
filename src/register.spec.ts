const mockAddonsRegister = jest.fn();
const mockAddonsAddPanel = jest.fn();

mockAddonsRegister.mockImplementation((_addonName: string, cb: any) => {
  cb("api");
});

jest.mock("@storybook/addons", () => ({
  default: {
    register: mockAddonsRegister,
    addPanel: mockAddonsAddPanel,
  },
}));

import { register, testingResetRegistered } from "./register";

// This test must come first due to internal module flag
describe("import test", () => {
  it("registers addon on module import", () => {
    expect(mockAddonsRegister.mock.calls).toHaveLength(1);
  });
});

describe("addon registration", () => {
  beforeEach(() => {
    mockAddonsRegister.mockReset();
    testingResetRegistered();
  });

  it("registers addon at most once", () => {
    expect(mockAddonsRegister.mock.calls).toHaveLength(0);
    register();
    expect(mockAddonsRegister.mock.calls).toHaveLength(1);
    register();
    expect(mockAddonsRegister.mock.calls).toHaveLength(1);
  });

  it("registers addon with name", () => {
    register();
    expect(mockAddonsRegister.mock.calls[0][0]).toMatchSnapshot();
  });

  it("adds panel", () => {
    register();

    // Panel name matches string from ./constants.ts
    expect(mockAddonsAddPanel.mock.calls[0]).toMatchSnapshot();

    // Creates Manager components, passing Storybook Addon channel and api
    expect(
      mockAddonsAddPanel.mock.calls[0][1].render.toString(),
    ).toMatchSnapshot();
  });
});
