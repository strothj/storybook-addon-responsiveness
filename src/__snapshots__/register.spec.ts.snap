// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addon registration adds panel 1`] = `
Array [
  "strothj/storybook-addon-responsiveness/panel",
  Object {
    "render": [Function],
    "title": "Responsiveness",
  },
]
`;

exports[`addon registration adds panel 2`] = `"function () {return React.createElement(Manager_1.default, { storybookChannel: addons_1.default.getChannel(), storybookAPI: api });}"`;

exports[`addon registration registers addon with name 1`] = `"strothj/storybook-addon-responsiveness"`;
